### ISSUE TITLE
Sprintf Rounding off Bug

### ISSUE THREAD

**Issue by modestlearner** [golang/go#36158]
### What version of Go are you using (`go version`)?

<pre>
$ go version
1.11.13

</pre>

### Does this issue reproduce with the latest release?
Yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
linux/amd64

</pre></details>

### What did you do?

fmt.Println(fmt.Sprintf("%.2f", 10.915)) 
fmt.Println(fmt.Sprintf("%.2f", 1.915)) 

### What did you expect to see?
10.91
1.92

### What did you see instead?
10.92
1.92

**Comment by ALTree** [comment]
Hi,

`10.915` is not representable as a IEE-754 float, so in practise the value is registered as 

```
10.91499999999999914734871708787977695465087890625
```

which is correctly rounded as 10.91.

Closing here, since this is working as intended.

**Comment by ALTree** [comment]
(you inverted  expected and observed btw, the observed is:

```
10.91
1.92
```
)

Anyway none of these number is exactly representable as an IEEE-754 float, so their real value is:
```
10.91499999999999914734871708787977695465087890625
1.91500000000000003552713678800500929355621337890625
```

which explains why they are rounded differently. Closing here.

**Comment by modestlearner** [comment]
So how can I achieve standard mathematical rounding up to 2 precision without using any external library?

**Comment by randall77** [comment]
I looked into this a bit more, and it is definitely a rathole.

TL;DR Go agrees with C. Java and PHP agree with the original poster.

C:
```
#include <stdio.h>
int main(int argc, char *argv[]) {
  double a = 1.915;
  double b = 10.915;
  printf("%.2f %.2f\n",a,b);
  printf("%.20f %.20f\n",a,b);
  printf("%llx %llx\n", *(long long*)(&a), *(long long*)(&b));
}
```
prints
```
1.92 10.91
1.91500000000000003553 10.91499999999999914735
3ffea3d70a3d70a4 4025d47ae147ae14
```

Java:
```
import java.lang.Double;
public class HelloWorld {
     public static void main(String []args) {
	 double a = 1.915;
	 double b = 10.915;
	 System.out.printf("%.2f %.2f\n", a, b);
	 System.out.printf("%.20f %.20f\n", a, b);
	 System.out.printf("%x %x\n", Double.doubleToLongBits(a), Double.doubleToLongBits(b));
     }
}
```
prints
```
1.92 10.92
1.91500000000000000000 10.91500000000000000000
3ffea3d70a3d70a4 4025d47ae147ae14
```

Note that the floating point bits are exactly the same in both cases, just the rounding is different.
Forget rounding for a second, the high-precision printing (%.20f) looks just wrong.
The results are the same if I initialize `a` and `b` using `Double.longBitsToDouble`.

**Comment by randall77** [comment]
It looks like Java rounds twice. It first constructs a string representation of the double value. Then it rounds using the string.
For the string representation it has some freedom - it chooses a string which will parse to exactly the double in question. There are many such strings. It computes just enough digits (16, I think) to ensure that the double is represented exactly. To 16 digits, the doubles in question end up at the original text values, 1.915 and 10.915.

In java if you print the result of 1/3 to 20 digits. you get `0.33333333333333330000`. In Go or C you get `0.33333333333333331483`. The latter is more exact, as the rounding only happens at the 20th digit, not the 16th digit.

**Comment by modestlearner** [comment]
@randall77 thanks for explaining this so comprehensively but again how can I achieve standard mathematical rounding in Go as I am able to do the same in PHP.

**Comment by randall77** [comment]
Well, you aren't really able to do that in PHP either. You happen to luck out in your specific example.
It is challenging to do exact math with floating point numbers. For instance, check the value of this expression: `0.1+0.1+0.1 == 0.3`. You'll note it is false in all the languages (if you appropriately disable compile-time simplification).
You probably want to use a big decimal implementation. Here's a start: https://godoc.org/?q=big+decimal

**Comment by proyb6** [comment]
While ANZ-Bank has a Decimals package which is still work-in-progress.

**Comment by randall77** [comment]
Java bug (5+ years old, still open): https://bugs.openjdk.java.net/browse/JDK-8066560
